Java/SegmentTree.java

```diff
@@ -0,0 +1,158 @@
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+/** A class to represent a Segment Tree.
+ */
+public class SegmentTree {
+    /**
+     * @param {List} arr   The segment tree will be built using this iterable
+     * @param {function} func  The function that will be used to built the tree
+     * @param {number} def     The default value with which the tree array is initialized
+     */
+    List<Integer> arr;
+    Operations func;
+    int def, n_leaves, len_arr;
+    List<Integer> tree;
+
+    public SegmentTree(List<Integer> arr, Operations func, int def) {
+        if (arr == null)
+            throw new NullPointerException("Argument arr should not be null");
+        this.arr = arr;
+        this.len_arr = this.arr.size();
+
+        if (func == null)
+            throw new NullPointerException("Argument func should not be null");
+        this.func = func;
+        this.def = def;
+
+        buildTree();
+    }
+
+    /**
+     * Function to return the next number which is a power of 2.
+     *
+     * @param n
+     * @returns number
+     */
+    private int next_power_of_2(int n) {
+        // Checks if n is already a power of 2
+        if ((n & n-1) == 0)
+            return n;
+        int val = 1;
+
+        // Loop will run until it give us the value which is greater than n.
+        while (val <= n)
+            // val will be left shifted by 1 in each iteration.
+            val <<= 1;
+
+        return val;
+    }
+
+    /**Function to build the segment tree using the function provided.
+     */
+    private void buildTree() {
+        // If number of leaves in a full binary tree is n,
+        // total number of internal nodes in the tree is n - 1.
+        // Hence, total number of nodes in the tree will be 2n - 1.
+        this.n_leaves = next_power_of_2(this.len_arr);
+        int n = this.n_leaves << 1;
+
+        // initializing all nodes of the tree to 0
+        this.tree = new ArrayList<>();
+        for (int i = 0; i < n; i++)
+            this.tree.add(this.def);
+
+        // copying array to leaves of the tree
+        for (int i = 0; i < this.len_arr; i++)
+            this.tree.set(this.len_arr+i, this.arr.get(i));
+
+        // updating all the nodes of the tree
+        for (int i = (int) (this.len_arr - 1); i > 0; i--)
```

Comment:
Unnecessary cast to int, when len_arr is already int

Reply:
Forgot to remove it, will do it asap